File: python/models/batch_control.py
"""BATCH-CONTROL-RECORD model from BCHCTL.cpy."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import List, Optional

from python.models.batch_constants import MAX_PREREQ, ProcessStatus, ReturnCode


@dataclass
class PrerequisiteJob:
    """BCT-PREREQ-JOBS occurrence (one of up to 10 entries).

    Field lengths from BCHCTL.cpy:
        BCT-PREREQ-NAME  PIC X(8)
        BCT-PREREQ-SEQ   PIC 9(4)
        BCT-PREREQ-RC    PIC S9(4) COMP
    """

    name: str = ""
    sequence: int = 0
    return_code: int = 0

    def __post_init__(self) -> None:
        self.name = self.name[:8]


@dataclass
class BatchControlRecord:
    """Mirror of BATCH-CONTROL-RECORD from BCHCTL.cpy.

    Field lengths:
        BCT-JOB-NAME       PIC X(8)
        BCT-PROCESS-DATE   PIC X(8)   (YYYYMMDD)
        BCT-SEQUENCE-NO    PIC 9(4)
        BCT-STATUS         PIC X(1)
        BCT-STEP-NAME      PIC X(8)
        BCT-PROGRAM-NAME   PIC X(8)
        BCT-START-TIME     PIC X(8)
        BCT-END-TIME       PIC X(8)
        BCT-PREREQ-COUNT   PIC 9(2) COMP
        BCT-PREREQ-JOBS    OCCURS 10 TIMES
        BCT-RETURN-CODE    PIC S9(4) COMP
        BCT-ERROR-DESC     PIC X(80)
        BCT-RESTART-COUNT  PIC 9(2) COMP
        BCT-ATTEMPT-TS     PIC X(26)
        BCT-COMPLETE-TS    PIC X(26)
        BCT-FILLER         PIC X(50)
        BCT-RECORDS-READ   (used by HISTLD00 2310-UPDATE-CHECKPOINT)
        BCT-RECORDS-WRITTEN
    """

    job_name: str = ""
    process_date: str = ""
    sequence_no: int = 0
    status: str = ProcessStatus.READY.value
    step_name: str = ""
    program_name: str = ""
    start_time: str = ""
    end_time: str = ""
    prereqs: List[PrerequisiteJob] = field(default_factory=list)
    return_code: int = ReturnCode.SUCCESS
    error_desc: str = ""
    restart_count: int = 0
    attempt_ts: str = ""
    complete_ts: str = ""
    records_read: int = 0
    records_written: int = 0

    def __post_init__(self) -> None:
        self.job_name = self.job_name[:8]
        self.process_date = self.process_date[:8]
        self.step_name = self.step_name[:8]
        self.program_name = self.program_name[:8]
        self.start_time = self.start_time[:8]
        self.end_time = self.end_time[:8]
        self.error_desc = self.error_desc[:80]
        self.attempt_ts = self.attempt_ts[:26]
        self.complete_ts = self.complete_ts[:26]
        if len(self.prereqs) > MAX_PREREQ:
            self.prereqs = self.prereqs[:MAX_PREREQ]

    @property
    def key(self) -> str:
        """Return the BCT-KEY composite (job + date + sequence)."""
        return f"{self.job_name:<8}{self.process_date:<8}{self.sequence_no:04d}"

    def mark_active(self, when: Optional[datetime] = None) -> None:
        """Transition status to ACTIVE and stamp BCT-ATTEMPT-TS."""
        when = when or datetime.now(timezone.utc)
        self.status = ProcessStatus.ACTIVE.value
        self.attempt_ts = _format_ts(when)
        self.start_time = when.strftime("%H:%M:%S")

    def mark_done(
        self,
        when: Optional[datetime] = None,
        return_code: int = ReturnCode.SUCCESS,
    ) -> None:
        """Transition status to DONE and stamp BCT-COMPLETE-TS."""
        when = when or datetime.now(timezone.utc)
        self.status = ProcessStatus.DONE.value
        self.complete_ts = _format_ts(when)
        self.end_time = when.strftime("%H:%M:%S")
        self.return_code = return_code

    def mark_error(
        self,
        error_desc: str,
        return_code: int = ReturnCode.ERROR,
        when: Optional[datetime] = None,
    ) -> None:
        """Transition status to ERROR and store the error description."""
        when = when or datetime.now(timezone.utc)
        self.status = ProcessStatus.ERROR.value
        self.complete_ts = _format_ts(when)
        self.end_time = when.strftime("%H:%M:%S")
        self.error_desc = error_desc[:80]
        self.return_code = return_code


def _format_ts(when: datetime) -> str:
    """Format a 26-character timestamp matching COBOL ``X(26)``."""
    base = when.strftime("%Y-%m-%d-%H.%M.%S.%f")
    return base[:26].ljust(26)

File: java/src/main/java/com/portfolio/domain/BatchControl.java
package com.portfolio.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import jakarta.persistence.EmbeddedId;
import jakarta.persistence.Entity;
import jakarta.persistence.Table;

import java.io.Serializable;
import java.util.Objects;

/**
 * Relational model of the VSAM KSDS batch control file (BCHCTL), the file
 * HISTLD00 opens I-O for checkpointing. Record layout:
 * {@code src/copybook/batch/BCHCTL.cpy} (01 BATCH-CONTROL-RECORD).
 *
 * <p>VSAM→table convention: primary key = COBOL RECORD KEY BCT-KEY
 * (BCT-JOB-NAME, BCT-PROCESS-DATE, BCT-SEQUENCE-NO). The OCCURS 10 TIMES
 * prerequisite table is not needed by the HISTLD00 slice and is deferred.
 */
@Entity
@Table(name = "VSAM_BCHCTL")
public class BatchControl {

    @EmbeddedId
    private Key key;

    /** BCT-STATUS PIC X(1) — R/A/W/D/E. */
    @Column(name = "STATUS", length = 1, nullable = false)
    private String status;

    /** BCT-STEP-NAME PIC X(8). */
    @Column(name = "STEP_NAME", length = 8)
    private String stepName;

    /** BCT-PROGRAM-NAME PIC X(8). */
    @Column(name = "PROGRAM_NAME", length = 8)
    private String programName;

    /** BCT-START-TIME PIC X(8). */
    @Column(name = "START_TIME", length = 8)
    private String startTime;

    /** BCT-END-TIME PIC X(8). */
    @Column(name = "END_TIME", length = 8)
    private String endTime;

    /** Records read counter maintained by HISTLD00 checkpointing (2310-UPDATE-CHECKPOINT). */
    @Column(name = "RECORDS_READ", nullable = false)
    private long recordsRead;

    /** Records written counter maintained by HISTLD00 checkpointing. */
    @Column(name = "RECORDS_WRITTEN", nullable = false)
    private long recordsWritten;

    /** BCT-RETURN-CODE PIC S9(4) COMP. */
    @Column(name = "RETURN_CODE", nullable = false)
    private int returnCode;

    /** BCT-ERROR-DESC PIC X(80). */
    @Column(name = "ERROR_DESC", length = 80)
    private String errorDesc;

    /** BCT-RESTART-COUNT PIC 9(2) COMP. */
    @Column(name = "RESTART_COUNT", nullable = false)
    private int restartCount;

    /** BCT-ATTEMPT-TS PIC X(26). */
    @Column(name = "ATTEMPT_TS", length = 26)
    private String attemptTimestamp;

    /** BCT-COMPLETE-TS PIC X(26). */
    @Column(name = "COMPLETE_TS", length = 26)
    private String completeTimestamp;

    /** Composite primary key = BCT-KEY (job name + process date + sequence no). */
    @Embeddable
    public static class Key implements Serializable {

        /** BCT-JOB-NAME PIC X(8). */
        @Column(name = "JOB_NAME", length = 8, nullable = false)
        private String jobName;

        /** BCT-PROCESS-DATE PIC X(8) — YYYYMMDD. */
        @Column(name = "PROCESS_DATE", length = 8, nullable = false)
        private String processDate;

        /** BCT-SEQUENCE-NO PIC 9(4). */
        @Column(name = "SEQUENCE_NO", nullable = false)
        private int sequenceNo;

        public Key() {}

        public Key(String jobName, String processDate, int sequenceNo) {
            this.jobName = jobName;
            this.processDate = processDate;
            this.sequenceNo = sequenceNo;
        }

        public String getJobName() { return jobName; }
        public void setJobName(String jobName) { this.jobName = jobName; }
        public String getProcessDate() { return processDate; }
        public void setProcessDate(String processDate) { this.processDate = processDate; }
        public int getSequenceNo() { return sequenceNo; }
        public void setSequenceNo(int sequenceNo) { this.sequenceNo = sequenceNo; }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof Key key)) return false;
            return sequenceNo == key.sequenceNo
                    && Objects.equals(jobName, key.jobName)
                    && Objects.equals(processDate, key.processDate);
        }

        @Override
        public int hashCode() {
            return Objects.hash(jobName, processDate, sequenceNo);
        }
    }

    public Key getKey() { return key; }
    public void setKey(Key key) { this.key = key; }
    public String getStatus() { return status; }
    public void setStatus(String status) { this.status = status; }
    public String getStepName() { return stepName; }
    public void setStepName(String stepName) { this.stepName = stepName; }
    public String getProgramName() { return programName; }
    public void setProgramName(String programName) { this.programName = programName; }
    public String getStartTime() { return startTime; }
    public void setStartTime(String startTime) { this.startTime = startTime; }
    public String getEndTime() { return endTime; }
    public void setEndTime(String endTime) { this.endTime = endTime; }
    public long getRecordsRead() { return recordsRead; }
    public void setRecordsRead(long recordsRead) { this.recordsRead = recordsRead; }
    public long getRecordsWritten() { return recordsWritten; }
    public void setRecordsWritten(long recordsWritten) { this.recordsWritten = recordsWritten; }
    public int getReturnCode() { return returnCode; }
    public void setReturnCode(int returnCode) { this.returnCode = returnCode; }
    public String getErrorDesc() { return errorDesc; }
    public void setErrorDesc(String errorDesc) { this.errorDesc = errorDesc; }
    public int getRestartCount() { return restartCount; }
    public void setRestartCount(int restartCount) { this.restartCount = restartCount; }
    public String getAttemptTimestamp() { return attemptTimestamp; }
    public void setAttemptTimestamp(String attemptTimestamp) { this.attemptTimestamp = attemptTimestamp; }
    public String getCompleteTimestamp() { return completeTimestamp; }
    public void setCompleteTimestamp(String completeTimestamp) { this.completeTimestamp = completeTimestamp; }
}
